File: leetcode/p01299/p01299.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def replaceElements(self, arr: List[int]) -> List[int]:
        maximum = arr[-1]
        arr[-1] = -1
        
        for n in range (len(arr)-2, -1, -1):
            if arr[n] > maximum:
                arr[n], maximum = maximum, arr[n]
            else:
                arr[n] = maximum
        
        return arr

if __name__ == '__main__':
    s = Solution()

    assert s.replaceElements([17,18,5,4,6,1]) == [18,6,6,6,1,-1]
    assert s.replaceElements([400]) == [-1]

File: leetcode/p01299/p01299.go
package p01299

// Time Complexity: O(n)
// Space Complexity: O(1)
func replaceElements(arr []int) []int {
	tail := len(arr) - 1
	maximum := arr[tail]
	arr[tail] = -1

	for i := tail - 1; i > -1; i-- {
		if arr[i] > maximum {
			arr[i], maximum = maximum, arr[i]
		} else {
			arr[i] = maximum
		}
	}

	return arr
}
